pallets/hyperdrive/src/types.rs

```diff
@@ -36,3 +43,99 @@ pub enum StateTransmitterUpdate<AccountId, BlockNumber> {
+}
+
+/// Verify Merkle Proof correctness versus given root and leaf hash.
+///
+/// The proof is NOT expected to contain leaf hash as the first
+/// element, but only all adjacent (blinded) nodes required to eventually by process of
+/// concatenating and hashing end up with given root hash.
+///
+/// The proof must not contain the root hash.
+pub fn verify_proof<'a, H, P>(root_hash: &'a H::Output, proof: P, leaf_hash: H::Output) -> bool
+where
+    H: Hash,
+    H::Output: PartialEq + AsRef<[u8]>,
+    P: IntoIterator<Item = StateProofNode<H::Output>>,
+{
+    let derived = derive_proof::<H, P>(proof, leaf_hash);
+    root_hash == &derived
+}
+
+pub(crate) fn derive_proof<'a, H, P>(proof: P, leaf_hash: H::Output) -> <H as Hash>::Output
+where
+    H: Hash,
+    H::Output: PartialEq + AsRef<[u8]>,
+    P: IntoIterator<Item = StateProofNode<H::Output>>,
+{
+    let hash_len = <H as sp_core::Hasher>::LENGTH;
+    let mut combined = vec![0_u8; hash_len * 2];
+    let computed = proof.into_iter().fold(leaf_hash, |a, b| {
+        match b {
+            StateProofNode::Right(h) => {
+                combined[..hash_len].copy_from_slice(&a.as_ref());
+                combined[hash_len..].copy_from_slice(&h.as_ref());
+            }
+            StateProofNode::Left(h) => {
+                combined[..hash_len].copy_from_slice(&h.as_ref());
+                combined[hash_len..].copy_from_slice(&a.as_ref());
+            }
+        }
+        let hash = <H as Hash>::hash(&combined);
+        #[cfg(feature = "debug_assertions")]
+        log::debug!(
+            "[verify_proof]: (a, b) {:?}, {:?} => {:?} ({:?}) hash",
+            array_bytes::bytes2hex("", &a.as_ref()),
+            array_bytes::bytes2hex("", &b.as_ref()),
+            array_bytes::bytes2hex("", &hash.as_ref()),
+            array_bytes::bytes2hex("", &combined.as_ref())
+        );
+        hash
+    });
+
+    computed
+}
+
+pub const MESSAGE_MAX_LENGTH: u32 = 5;
+pub type Message = BoundedVec<u8, ConstU32<MESSAGE_MAX_LENGTH>>;
+
+#[derive(RuntimeDebug, Encode, Decode, TypeInfo, Clone, PartialEq, EnumString)]
+pub enum RawAction {
+    #[strum(serialize = "REGISTER_JOB")]
+    RegisterJob,
+}
+
+#[derive(RuntimeDebug, Encode, Decode, TypeInfo, Clone, PartialEq)]
+pub enum ParsedAction<AccountId, Extra> {
+    RegisterJob(JobId<AccountId>, JobRegistration<AccountId, Extra>),
+}
+
+pub type JobRegistrationFor<T> =
+    JobRegistration<<T as frame_system::Config>::AccountId, <T as Config>::RegistrationExtra>;
+
+pub trait MessageParser<AccountId, Extra> {
+    type Error;
+
+    fn parse(encoded: &[u8]) -> Result<ParsedAction<AccountId, Extra>, Self::Error>;
+}
+
+pub trait ActionExecutor<AccountId, Extra> {
+    fn execute(action: ParsedAction<AccountId, Extra>) -> DispatchResultWithPostInfo;
+}
+
+impl<AccountId, Extra> ActionExecutor<AccountId, Extra> for () {
+    fn execute(_: ParsedAction<AccountId, Extra>) -> DispatchResultWithPostInfo {
+        Ok(().into())
+    }
+}
```

Comment:
`unit` implementation does not make sense here. I'm fine if this gets moved to `mock.rs`
```suggestion
```